src/session.rs

```diff
@@ -0,0 +1,340 @@
+                    secure: cookie.secure(),
+                },
+            );
+        }
+    }
+
+    pub fn persist(&self) -> Result<()> {
+        if !self.path.exists() || !self.read_only {
+            if let Some(parent_path) = self.path.parent() {
+                fs::create_dir_all(parent_path)?;
+            }
+            let mut session_file = fs::File::create(&self.path)?;
+            let formatter = serde_json::ser::PrettyFormatter::with_indent(b"    ");
+            let mut ser = serde_json::Serializer::with_formatter(&mut session_file, formatter);
+            self.content.serialize(&mut ser)?;
+            session_file.write_all(b"\n")?;
+        }
+        Ok(())
+    }
+}
+
+fn is_path(value: &OsString) -> bool {
+    value.to_string_lossy().contains(std::path::is_separator)
+}
+
+fn config_dir() -> Option<PathBuf> {
+    if test_mode() {
+        std::env::var_os("XH_TEST_CONFIG_DIR").map(Into::into)
+    } else {
+        dirs::config_dir()
+    }
+}
+
+fn xh_version() -> String {
+    if test_mode() {
+        "0.0.0".into()
+    } else {
+        env!("CARGO_PKG_VERSION").into()
+    }
+}
+
+fn path_from_url(url: &Url) -> Result<String> {
+    match (url.host_str(), url.port()) {
+        (Some("."), _) | (Some(".."), _) | (None, _) => {
+            Err(anyhow!("couldn't extract host from url"))
+        }
+        (Some(host), Some(port)) => Ok(format!("{}_{}", host, port)),
+        (Some(host), None) => Ok(host.into()),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::utils::random_string;
+    use anyhow::Result;
+
+    #[test]
+    fn can_read_httpie_session_file() -> Result<()> {
+        let mut path_to_session = std::env::temp_dir();
+        let file_name = random_string();
+        path_to_session.push(file_name);
+        fs::write(
+            &path_to_session,
+            indoc::indoc! {r#"
+                {
+                    "__meta__": {
+                        "about": "HTTPie session file",
+                        "help": "https://httpie.org/doc#sessions",
+                        "httpie": "2.3.0"
+                    },
+                    "auth": {
+                        "password": null,
+                        "type": null,
+                        "username": null
+                    },
+                    "cookies": {
+                        "__cfduid": {
+                            "expires": 1620239688,
+                            "path": "/",
+                            "secure": false,
+                            "value": "d090ada9c629fc7b8bbc6dba3dde1149d1617647688"
+                        }
+                    },
+                    "headers": {
+                        "authorization": "bearer hello"
+                    }
+                }
+            "#},
+        )?;
+
+        Session::load_session(
+            &Url::parse("http://localhost")?,
+            path_to_session.into(),
+            false,
+        )?;
+        Ok(())
```

Comment:
Can you test that it actually loads the cookie and header?

Idem for `can_read_xh_session_file`.